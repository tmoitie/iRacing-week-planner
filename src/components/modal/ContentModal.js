// @flow

import * as React from 'react';
import difference from 'lodash.difference';
import isEqual from 'lodash.isequal';
import { useTranslation } from 'react-i18next';
import toggleIdInCollection from '../../lib/toggleIdInCollection';
import Modal from './Modal';
import Checkbox from '../Checkbox';
import FavouriteStarButton from '../FavouriteStarButton';
import styles from '../../styles/main.module.scss';
import contentModalStyles from './styles/contentModal.module.scss';

type Props = {
  id: string,
  isOpen: boolean,
  onClose: () => void,
  title: string,
  save: (Array<number>) => void,
  content: Array<{skuname?: string, name?: string}>,
  idField: string,
  defaultContent: Array<number>,
  ownedContent: Array<number>,
  favourites: Array<number>,
  saveFavourites: (Array<number>) => void,
  typeFilter: {
    key: string,
    oval: string,
    road: string,
    dirtOval: string,
    dirtRoad: string,
  } | {
    key: string,
    oval: string,
    formulaCar: string,
    sportsCar: string,
    dirtOval: string,
    dirtRoad: string,
  }
};

export default function ContentModal({
  id, isOpen, onClose, title, save, content, idField, defaultContent, ownedContent, favourites,
  saveFavourites, typeFilter,
}: Props) {
  const { t } = useTranslation();

  const toggleAllContent = (newValue) => {
    save(newValue ? content.map((item) => item[idField]) : defaultContent);
  };

  const getAllIdsOfType = (type) => {
    const { key } = typeFilter;
    const value = typeFilter[type];
    return content.filter((item) => isEqual(item[key], value)).map((item) => item[idField]);
  };

  const toggleAllTypeInCollection = (type, newState, currentCollection, passedDefaultContent) => {
    const allIdsOfType = getAllIdsOfType(type);

    if (newState) {
      const addValues = difference(allIdsOfType, currentCollection);

      return [...currentCollection, ...addValues];
    }

    return difference(currentCollection, difference(allIdsOfType, passedDefaultContent));
  };

  const setAllTypeContent = (type, newValue) => {
    save(toggleAllTypeInCollection(type, newValue, ownedContent, defaultContent));
  };

  const setAllTypeFavourites = (type, newState) => {
    saveFavourites(toggleAllTypeInCollection(type, newState, favourites, []));
  };

  const allTypeContentChecked = (type) => {
    const allIdsOfType = getAllIdsOfType(type);
    return difference(allIdsOfType, ownedContent).length === 0;
  };

  const allTypeFavouritesChecked = (type) => {
    const allIdsOfType = getAllIdsOfType(type);
    return difference(allIdsOfType, favourites).length === 0;
  };

  const toggleContent = (contentId, newValue) => {
    const newContent = toggleIdInCollection(ownedContent, contentId, newValue);
    save(newContent);
  };

  const toggleFavourite = (contentId, newState) => {
    const newFavourites = toggleIdInCollection(favourites, contentId, newState);
    saveFavourites(newFavourites);
  };

  const toggleAllFavourites = (newState) => {
    saveFavourites(newState ? content.map((item) => item[idField]) : []);
  };

  return (
    <Modal isOpen={isOpen} onClose={onClose} title={title} doneAction={onClose} id={id}>
      <div className={styles['container-fluid']}>
        <p>{t('Use the checkbox to set the content you own, and use the star to set your favourite content.')}</p>
        <div className={styles.row}>
          <div className={`${contentModalStyles.checkboxContainer} ${styles['col-sm-3']}`}>
            <div>
              <Checkbox
                id={`${id}-select-all`}
                checked={ownedContent.length === content.length}
                onChange={toggleAllContent}
              >
                {t('Select all')}
              </Checkbox>
            </div>

            <div className={contentModalStyles.favourite}>
              <FavouriteStarButton
                id={`${id}-favourite-all`}
                enabled={favourites.length === content.length}
                onClick={toggleAllFavourites}
              />
            </div>
          </div>
          <div className={`${contentModalStyles.checkboxContainer} ${styles['col-sm-3']}`}>
            <div>
              <Checkbox
                id={`${id}-select-oval`}
                checked={allTypeContentChecked('oval')}
                onChange={(newValue) => setAllTypeContent('oval', newValue)}
              >
                {t('Select all oval')}
              </Checkbox>
            </div>

            <div className={contentModalStyles.favourite}>
              <FavouriteStarButton
                id={`${id}-favourite-oval`}
                enabled={allTypeFavouritesChecked('oval')}
                onClick={(newValue) => setAllTypeFavourites('oval', newValue)}
              />
            </div>
          </div>

          {typeFilter.road ? (
            <div className={`${contentModalStyles.checkboxContainer} ${styles['col-sm-3']}`}>
              <div>
                <Checkbox
                  id={`${id}-select-road`}
                  checked={allTypeContentChecked('road')}
                  onChange={(newValue) => setAllTypeContent('road', newValue)}
                >
                  {t('Select all road')}
                </Checkbox>
              </div>

              <div className={contentModalStyles.favourite}>
                <FavouriteStarButton
                  id={`${id}-favourite-road`}
                  enabled={allTypeFavouritesChecked('road')}
                  onClick={(newValue) => setAllTypeFavourites('road', newValue)}
                />
              </div>
            </div>
          ) : (
            <>
              <div className={`${contentModalStyles.checkboxContainer} ${styles['col-sm-3']}`}>
                <div>
                  <Checkbox
                    id={`${id}-select-sports-cars`}
                    checked={allTypeContentChecked('sportsCar')}
                    onChange={(newValue) => setAllTypeContent('sportsCar', newValue)}
                  >
                    {t('Select all sports cars')}
                  </Checkbox>
                </div>

                <div className={contentModalStyles.favourite}>
                  <FavouriteStarButton
                    id={`${id}-favourite-sports-cars`}
                    enabled={allTypeFavouritesChecked('sportsCar')}
                    onClick={(newValue) => setAllTypeFavourites('sportsCar', newValue)}
                  />
                </div>
              </div>
              <div className={`${contentModalStyles.checkboxContainer} ${styles['col-sm-3']}`}>
                <div>
                  <Checkbox
                    id={`${id}-select-formula-cars`}
                    checked={allTypeContentChecked('formulaCar')}
                    onChange={(newValue) => setAllTypeContent('formulaCar', newValue)}
                  >
                    {t('Select all formula cars')}
                  </Checkbox>
                </div>

                <div className={contentModalStyles.favourite}>
                  <FavouriteStarButton
                    id={`${id}-favourite-formula-cars`}
                    enabled={allTypeFavouritesChecked('formulaCar')}
                    onClick={(newValue) => setAllTypeFavourites('formulaCar', newValue)}
                  />
                </div>
              </div>
            </>
          )}

          <div className={`${contentModalStyles.checkboxContainer} ${styles['col-sm-3']}`}>
            <div>
              <Checkbox
                id={`${id}-select-dirt-road`}
                checked={allTypeContentChecked('dirtRoad')}
                onChange={(newValue) => setAllTypeContent('dirtRoad', newValue)}
              >
                {t('Select all dirt road')}
              </Checkbox>
            </div>

            <div className={contentModalStyles.favourite}>
              <FavouriteStarButton
                id={`${id}-favourite-dirt-road`}
                enabled={allTypeFavouritesChecked('dirtRoad')}
                onClick={(newValue) => setAllTypeFavourites('dirtRoad', newValue)}
              />
            </div>
          </div>

          <div className={`${contentModalStyles.checkboxContainer} ${styles['col-sm-3']}`}>
            <div>
              <Checkbox
                id={`${id}-select-dirt-oval`}
                checked={allTypeContentChecked('dirtOval')}
                onChange={(newValue) => setAllTypeContent('dirtOval', newValue)}
              >
                {t('Select all dirt oval')}
              </Checkbox>
            </div>

            <div className={contentModalStyles.favourite}>
              <FavouriteStarButton
                id={`${id}-favourite-dirt-oval`}
                enabled={allTypeFavouritesChecked('dirtOval')}
                onClick={(newValue) => setAllTypeFavourites('dirtOval', newValue)}
              />
            </div>
          </div>
        </div>

        <hr />

        <div className={styles.row}>
          {content.map((item) => (
            <div
              className={`${contentModalStyles.checkboxContainer} ${styles['col-md-6']}`}
              key={item[idField]}
            >
              <div>
                <Checkbox
                  id={`${id}-select-item-${item[idField]}`}
                  disabled={
                    ownedContent.indexOf(item[idField]) !== -1 && defaultContent.indexOf(item[idField]) !== -1
                  }
                  checked={ownedContent.indexOf(item[idField]) !== -1}
                  onChange={(newValue) => toggleContent(item[idField], newValue)}
                >
                  {item.skuname ? t(item.skuname) : t(item.name)}
                </Checkbox>
              </div>

              <div className={contentModalStyles.favourite}>
                <FavouriteStarButton
                  id={`${id}-favourite-item-${item[idField]}`}
                  enabled={favourites.indexOf(item[idField]) !== -1}
                  onClick={(newValue) => toggleFavourite(item[idField], newValue)}
                />
              </div>
            </div>
          ))}
        </div>
      </div>
    </Modal>
  );
}
