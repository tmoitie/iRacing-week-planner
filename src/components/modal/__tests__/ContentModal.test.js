import { describe, test, expect } from '@jest/globals';
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ContentModal from '../ContentModal';

describe('components/modal/ContentModal', () => {
  test('renders positive interactions correctly', async () => {
    const user = userEvent.setup();
    const save = jest.fn();
    const saveFavourites = jest.fn();
    const { container } = render(
      <ContentModal
        id="cars"
        title="Cars Modal"
        isOpen
        onClose={() => {}}
        save={save}
        content={[
          { id: 123, skuname: 'Ford', type: 'oval' },
          { id: 234, skuname: 'Toyota', type: 'oval' },
          { id: 456, name: 'Mazda', type: 'road' },
          { id: 645, name: 'McLaren', type: 'road' },
          { id: 999, name: 'Ferrari', type: 'road' },
        ]}
        ownedContent={[999]}
        defaultContent={[456]}
        idField="id"
        typeFilter={{
          key: 'type',
          oval: 'oval',
          road: 'road',
        }}
        favourites={[456]}
        saveFavourites={saveFavourites}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();

    await user.click(screen.getByLabelText('Select all oval'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 234, 999]));
    save.mockClear();

    await user.click(screen.getByLabelText('Select all road'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([456, 645, 999]));
    save.mockClear();

    await user.click(screen.getByLabelText('Select all'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 234, 456, 645, 999]));
    save.mockClear();

    await user.click(screen.getByLabelText('Toyota'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([234, 999]));
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-oval'));
    expect(saveFavourites).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 234, 456]));
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-road'));
    expect(saveFavourites).toHaveBeenCalledWith(expect.toIncludeSameMembers([456, 645, 999]));
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-all'));
    expect(saveFavourites).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 234, 456, 645, 999]));
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-item-123'));
    expect(saveFavourites).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 456]));
    save.mockClear();
  });

  test('renders negative interactions correctly', async () => {
    const user = userEvent.setup();
    const save = jest.fn();
    const saveFavourites = jest.fn();
    const { container } = render(
      <ContentModal
        id="cars"
        title="Cars Modal"
        isOpen
        onClose={() => {}}
        save={save}
        content={[
          { id: 123, skuname: 'Ford', type: 'oval' },
          { id: 234, skuname: 'Toyota', type: 'oval' },
          { id: 456, name: 'Mazda', type: 'road' },
          { id: 645, name: 'McLaren', type: 'road' },
          { id: 999, name: 'Ferrari', type: 'road' },
        ]}
        ownedContent={[123, 234, 456, 645, 999]}
        defaultContent={[456]}
        idField="id"
        typeFilter={{
          key: 'type',
          oval: 'oval',
          road: 'road',
        }}
        favourites={[123, 234, 456, 645, 999]}
        saveFavourites={saveFavourites}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();

    await user.click(screen.getByLabelText('Select all oval'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([456, 645, 999]));
    save.mockClear();

    await user.click(screen.getByLabelText('Select all road'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 234, 456]));
    save.mockClear();

    await user.click(screen.getByLabelText('Select all'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([456]));
    save.mockClear();

    await user.click(screen.getByLabelText('Toyota'));
    expect(save).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 456, 645, 999]));
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-oval'));
    expect(saveFavourites).toHaveBeenCalledWith(expect.toIncludeSameMembers([456, 645, 999]));
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-road'));
    expect(saveFavourites).toHaveBeenCalledWith(expect.toIncludeSameMembers([123, 234]));
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-all'));
    expect(saveFavourites).toHaveBeenCalledWith([]);
    save.mockClear();

    await user.click(screen.getByTestId('cars-favourite-item-123'));
    expect(saveFavourites).toHaveBeenCalledWith(expect.toIncludeSameMembers([234, 456, 645, 999]));
    save.mockClear();
  });
});
