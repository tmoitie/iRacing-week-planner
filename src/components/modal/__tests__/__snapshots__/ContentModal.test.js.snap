// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/ContentModal renders negative interactions correctly 1`] = `null`;

exports[`components/modal/ContentModal renders positive interactions correctly 1`] = `null`;
